taichi/backends/opengl/opengl_api.cpp

```diff
@@ -227,15 +231,132 @@ void initialize_opengl() {
   }
 }
 
-CompiledGLSL::CompiledGLSL(const std::string &source)
-    : glsl(std::make_unique<GLProgram>(GLShader(source))) {
-  glsl->link();
-}
+struct CompiledKernel {
+  std::string kernel_name;
+  std::unique_ptr<GLProgram> glsl;
+  int num_groups;
+  UsedFeature used;
+
+  // disscussion:
+  // https://github.com/taichi-dev/taichi/pull/696#issuecomment-609332527
+  CompiledKernel(CompiledKernel &&) = default;
+  CompiledKernel &operator=(CompiledKernel &&) = default;
+
+  explicit CompiledKernel(const std::string &kernel_name_,
+                          const std::string &kernel_source_code,
+                          int num_groups_,
+                          const UsedFeature &used_)
+      : kernel_name(kernel_name_),
+        glsl(std::make_unique<GLProgram>(GLShader(kernel_source_code))),
+        num_groups(num_groups_),
+        used(used_) {
+    glsl->link();
+    TI_DEBUG("source of kernel [{}] * {}:\n{}", kernel_name, num_groups,
+             kernel_source_code);
+#ifdef _GLSL_DEBUG
+    std::ofstream(fmt::format("/tmp/{}.comp", kernel_name))
+        .write(kernel_source_code.c_str(), kernel_source_code.size());
+#endif
+  }
+
+  void launch() const {
+    // TI_PERF();
+    glsl->use();
+
+    // https://www.khronos.org/opengl/wiki/Compute_Shader
+    // https://community.arm.com/developer/tools-software/graphics/b/blog/posts/get-started-with-compute-shaders
+    // https://www.khronos.org/assets/uploads/developers/library/2014-siggraph-bof/KITE-BOF_Aug14.pdf
+    //
+    // `glDispatchCompute(X, Y, Z)`   - the X*Y*Z  == `Blocks`   in CUDA
+    // `layout(local_size_x = X) in;` - the X      == `Threads`  in CUDA
+    //
+    glDispatchCompute(num_groups, 1, 1);
+    // TI_PERF(kernel_name.c_str(), kernel_name.size(), 107);
+  }
+};
+
+struct CompiledProgram::Impl {
+  std::vector<std::unique_ptr<CompiledKernel>> kernels;
+  int arg_count;
+  std::map<int, size_t> ext_arr_map;
+  size_t gtmp_size;
+
+  Impl(Kernel *kernel, size_t gtmp_size) : gtmp_size(gtmp_size) {
+    arg_count = kernel->args.size();
+    for (int i = 0; i < arg_count; i++) {
+      if (kernel->args[i].is_nparray) {
+        ext_arr_map[i] = kernel->args[i].size;
+      }
+    }
+  }
+
+  void add(const std::string &kernel_name,
+           const std::string &kernel_source_code,
+           int num_groups,
+           const UsedFeature &used) {
+    kernels.push_back(std::make_unique<CompiledKernel>(
+        kernel_name, kernel_source_code, num_groups, used));
```

Comment:
How about to use std::forward?

Reply:
you mean `std::forward<CompiledKernel>(std::make_unique<CompiledKernel>(...))`? That shouldn't be necessary, because `make_unique` is already an rvalue..

Reply:
No, I mean, CompiledKernel(forward<Args>(...)), like the python(*args, **kwargs), so that the args of CompiledKernel could change, sorry about the ambious.

Reply:
Sorry but this doesn't make sense. `std::forward`'s is for *perfect forwarding*. It's mostly used inside function template that takes in universal references (`T&&`), and has no similarity to python's  variable length functions.  <Effective Modern C++> would be a good read to understand what it does.

If you are concerned with having too many args passed to `CompiledKernel`, just simply create a `CompiledKernel::Params`, which contains all the args and will be passed to `CompiledKernel`'s constructor. This is the builder's pattern. E.g. 

https://github.com/taichi-dev/taichi/blob/d28533c503a1e4fd1101b549d63c6fb540618be2/taichi/backends/metal/kernel_manager.h#L26-L34

BTW, if there's no other comment, I'd hope to merge this asap. Sorry for the rush, but you probably have some local works on OpenGL as well.  The sooner we merge, the less conflicts this PR would cause...

Reply:
Thank for your hard working! Please merge this now, will add `CompiledKernel::Params` in another PR.